framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js

```diff
@@ -60,12 +63,142 @@ class BlockSynchronizationMechanism {
+			const lastCommonBlock = await this._revertToLastCommonBlock(
+				receivedBlock,
+				bestPeer,
+			);
+			await this._requestAndApplyBlocksToCurrentChain(
+				receivedBlock,
+				lastCommonBlock,
+				bestPeer,
+			);
 		} finally {
 			this.active = false;
 		}
 	}
 
+	/**
+	 * Request blocks from `fromID` ID to `toID` ID from an specific peer `peer`
+	 *
+	 * @param {object} peer - The peer to target
+	 * @param {string} fromID - The starting block ID to fetch from
+	 * @param {string} toID - The ending block ID
+	 * @return {Promise<Array<object>>}
+	 * @private
+	 */
+	async _requestBlocksWithinIDs(peer, fromID, toID) {
+		const maxFailedAttempts = 10; // TODO: Probably expose this to the configuration layer?
+		const blocks = [];
+		let failedAttempts = 0; // Failed attempt === the peer doesn't return any block or there is a network failure (no response or takes too long to answer)
+		let lastFetchedID = fromID;
+
+		while (failedAttempts < maxFailedAttempts) {
+			const { data } = await this.channel.invoke('network:requestFromPeer', {
+				procedure: 'getBlocksFromId',
+				peerId: peer.id,
+				data: {
+					blockId: lastFetchedID,
+				},
+			}); // Note that the block matching lastFetchedID is not returned but only higher blocks.
+
+			if (data) {
+				blocks.push(...data); // `data` is an array of blocks.
+				lastFetchedID = data.slice(-1).pop().id;
+				const index = blocks.findIndex(block => block.id === toID);
+				if (index > -1) {
+					return blocks.splice(0, index + 1); // Removes unwanted extra blocks
+				}
+			} else {
+				failedAttempts += 1; // It's only considered a failed attempt if the target peer doesn't provide any blocks on a single request
+			}
+		}
+
+		return blocks;
+	}
+
+	/**
+	 * Applies a set of blocks on top of the current chain
+	 * @param {Array<object>} blocks - An array of full blocks
+	 * @return {Promise<void>}
+	 * @private
+	 */
+	async _applyBlocksToCurrentChain(blocks) {
+		for (const block of blocks) {
+			await this.processorModule.process(addBlockProperties(block));
+		}
+	}
+
+	/**
+	 * Requests blocks from startingBlockID to an specific peer until endingBlockID
+	 * is met and applies them on top of the current chain.
+	 * @param receivedBlock
+	 * @param lastCommonBlock
+	 * @param {Object} peer - The peer to target
+	 * @return {Promise<void | boolean>}
+	 * @private
+	 */
+	async _requestAndApplyBlocksToCurrentChain(
+		receivedBlock,
+		lastCommonBlock,
+		peer,
+	) {
+		this.logger.debug(
+			{ peer, fromID: lastCommonBlock.id, toID: receivedBlock.id },
+			'Requesting blocks within IDs from peer',
+		);
+
+		const listOfFullBlocks = await this._requestBlocksWithinIDs(
+			peer,
+			lastCommonBlock.id,
+			receivedBlock.id,
+		);
+		const tipBeforeApplying = cloneDeep(this.blocks.lastBlock);
+
+		if (!listOfFullBlocks.length) {
+			return this._applyPenaltyAndRestartSync(
+				peer,
+				receivedBlock,
+				new Error("Peer didn't return any block after calling getBlocksFromId"),
```

Comment:
Instead of an error object, how about defining error codes for the penalty?
example: PEER_PENALTY_EMPTY_BLOCKS = `code here`